*: Remove hooks

wking opened the pull request:
This drops all instances (outside the ChangeLog) turned up by
case-insensitive searches for 'hook'.

The desire to drop hooks has been around [for a while now][1], but it
took a while to land a create/start split.  By the time that split did
land, dropping hooks was still [part of its motivation][2].  Hooks are
hard to [support robustly][3], and after the create/start split,
callers can do whatever they like at those times without having to [go
through the runtime][4].  There is still a use-case for folks who
prefer the old all-in-one UX, but we can support those folks with a
[higher-level wrapper built on hookless create/start primatives][5].
There was some last-minute discussion of [pre-pivot mount propagation
needing pre-start hooks][6], but that use case can be addressed by
[manipulating the mounts array][7].  With those arguments in place, the
consensus at today's meeting seemed in favor of [removing hooks from
the spec][8].

Fixes #473.
Fixes #472.
Fixes #427.
Fixes #395.
Fixes #20.

[1]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2015/opencontainers.2015-10-28-17.02.log.html#l-71
[2]: https://github.com/opencontainers/runtime-spec/pull/384#issue-147984345
[3]: https://github.com/opencontainers/runtime-spec/pull/265
[4]: http://ircbot.wl.linuxfoundation.org/eavesdrop/%23opencontainers/%23opencontainers.2016-03-24.log.html#t2016-03-24T18:56:15
[5]: https://groups.google.com/a/opencontainers.org/d/msg/dev/Y7p6YW8zr4s/OVaAI_WDBAAJ
[6]: https://github.com/opencontainers/runtime-spec/pull/384#issuecomment-220038190
[7]: https://github.com/opencontainers/runtime-spec/pull/384#issuecomment-220740126
[8]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2016/opencontainers.2016-06-01-17.01.log.html#l-83

mrunalp commented:
I'm not sure about this and have started a thread on the mailing list.

wking commented:
On Thu, Jun 02, 2016 at 08:38:24AM -0700, Mrunal Patel wrote:
> I'm not sure about this and have started a thread on the mailing
> list.

Link to thread [1].

[1]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/Y7p6YW8zr4s
     Subject: Hooks and all-in-one operation
     Date: Wed, 1 Jun 2016 11:38:27 -0700
     Message-ID: <CANEZBD69Q4VV4UoBodWjtAR1PrJ1OysaMUHgA5cW_aGj2ZEhLQ@mail.gmail.com>

jamesodhunt commented:
This isn't strictly relevant to the OCI spec, but hooks are currently used by docker to configure the container network. If they go away, it's unclear to me how networking will be handled, so this change could well break docker.

Maybe hooks should be retained until there is some sort of OCI network spec since, like annotations, they are a useful facility to handle arbitrary container scenarios.

wking commented:
On Fri, Jun 17, 2016 at 01:53:59AM -0700, James Hunt wrote:
> If they go away, it's unclear to me how networking will be handled…

We should keep discussion in @mrunalp's thread [1] to avoid
fragmentation.  I think [2] addresses your concerns, but if not, post
to the list thread with why ;).

[1]: https://github.com/opencontainers/runtime-spec/pull/483#issuecomment-223348774
[2]: https://groups.google.com/a/opencontainers.org/d/msg/dev/Y7p6YW8zr4s/NtiVw9p2AwAJ
     Subject: Re: Hooks and all-in-one operation
     Date: Thu, 9 Jun 2016 14:50:49 -0700
     Message-ID: <20160609215049.GK8801@odin.tremily.us>

wking commented:
On Fri, Jun 17, 2016 at 09:09:31AM -0700, W. Trevor King wrote:
> We should keep discussion in @mrunalp's thread [1] to avoid
> fragmentation.
> …
> [1]: https://github.com/opencontainers/runtime-spec/pull/483#issuecomment-223348774

I posted to the thread on Tuesday asking for guidance now that the
consensus seems to have swung back around to “drop hooks”, but haven't
heard back.  Rebasing this into master turned out to be not
particularly bad, so I just went ahead and did it with ea41a14 →
0113d7a.

wking commented:
So consensus on *today's* meeting was to keep hooks [1].  To square
that against the DockerCon decision to drop them [2], the current
consensus on facts seem to be:

a. The create/start split lets you do any hook-like things you like
   without hooks.
b. There are significant existing users who are using hooks as they
   stand.
c. Specifying hooks will take some effort to get right.
d. The OCI will likely manage the hook spec if that gets punted to a
   higher layer.
e. Nobody wants folks relying on unspecified behavior for important
   stuff like network setup.
f. Once you put something into 1.0, you need to wait until 2.0 to take
   it back out.

Possible approaches:

i. Include both hooks and the create/start split in this spec.
ii. Require the create/start split in this spec, and define hooks but
    make them a separate protocol [3].
iii. Punt the hook specification to a higher layer (possibly still
     within the OCI, see (d)), and have a tool that reads a config
     containing hook information, wraps the base runtime, and calls
     the hook commands at the appropriate times.

The DockerCon decision seemed to revolve around (a) and (f), arguing
for removing hooks as a conservative step, and possibly adding them
back after 1.0 if there was a demonstrated need (although in the light
of (a), I'm not sure what that would have looked like).

Today's decision seemed to revolve around (b), (d), and (e).  Because
of (d), punting doesn't really effect OCI time, so (c) is less
important for the OCI.  It does slightly raise the cost for base
runtime implementers (who now have to handle hooks), but they aren't
that complicated to implement (they're more complicated to
specify/test ;).

The distinction between these three choices is small because of (d),
and the only real issue with carving hooks out into a separate
protocol/spec or not is (f).  The consensus on today's call was that
the backwards-compatibility inertia from (b) was sufficient that (f)
was unlikely to be an issue.

[1]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2016/opencontainers.2016-08-17-20.59.log.html#l-61
[2]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2016/opencontainers.2016-08-03-21.01.log.html#l-21
[3]: https://github.com/opencontainers/runtime-spec/pull/527
